molgenis-navigator/src/main/java/org/molgenis/navigator/copy/service/CopyServiceImpl.java

```diff
@@ -0,0 +1,116 @@
+package org.molgenis.navigator.copy.service;
+
+import static com.google.common.collect.Iterables.size;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.molgenis.data.UnknownPackageException;
+import org.molgenis.data.meta.MetaDataService;
+import org.molgenis.data.meta.model.EntityType;
+import org.molgenis.data.meta.model.Package;
+import org.molgenis.data.util.PackageUtils.PackageTreeTraverser;
+import org.molgenis.i18n.CodedRuntimeException;
+import org.molgenis.i18n.ContextMessageSource;
+import org.molgenis.jobs.Progress;
+import org.molgenis.navigator.copy.exception.CopyFailedException;
+import org.molgenis.navigator.model.ResourceIdentifier;
+import org.molgenis.navigator.util.ResourceCollection;
+import org.molgenis.navigator.util.ResourceCollector;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.transaction.annotation.Transactional;
+
+@SuppressWarnings("unused")
+@Component
+public class CopyServiceImpl implements CopyService {
+
+  private final ResourceCollector resourceCollector;
+  private final MetaDataService metadataService;
+  private final PackageCopier packageCopier;
+  private final EntityTypeCopier entityTypeCopier;
+  private final ContextMessageSource contextMessageSource;
+
+  CopyServiceImpl(
+      ResourceCollector resourceCollector,
+      MetaDataService metadataService,
+      PackageCopier packageCopier,
+      EntityTypeCopier entityTypeCopier,
+      ContextMessageSource contextMessageSource) {
+    this.resourceCollector = requireNonNull(resourceCollector);
+    this.metadataService = requireNonNull(metadataService);
+    this.packageCopier = requireNonNull(packageCopier);
+    this.entityTypeCopier = requireNonNull(entityTypeCopier);
+    this.contextMessageSource = requireNonNull(contextMessageSource);
+  }
+
+  @Override
+  @Transactional(isolation = Isolation.SERIALIZABLE)
+  public Void copy(List<ResourceIdentifier> resources, String targetPackageId, Progress progress) {
+    try {
+      ResourceCollection resourceCollection = resourceCollector.get(resources);
+      Package targetPackage = getPackage(targetPackageId);
+      CopyState state = CopyState.create(targetPackage, progress);
+
+      copyResources(resourceCollection, state);
+
+    } catch (CodedRuntimeException exception) {
```

Comment:
Wrapping the CodedRuntimeException in a CopyFailedException will hide the error code of the wrapped exception. How about e.g. ```RecursiveCopyException extends CopyFailedException``` (similar to the ```IOException``` hierarchy).